### PR TITLE
Added proper error handling around sync resources

### LINKED ISSUE #1104: Completion of error handling
Would you like to add more error handling for return values from functions like the following?
- [pthread_cond_init](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cond_init.html "Initialize a condition variable.") ⇒ [CUDTSocket::construct()](https://github.com/Haivision/srt/blob/eee6628078b2367a80f0019518858d51125b1bf7/srtcore/api.cpp#L84)
- [pthread_mutex_init](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_init.html "Initialize a mutex.") ⇒ [srt::sync::Mutex constructor](https://github.com/Haivision/srt/blob/a080a3f6393cfc9216f9514ba5b2d4db7aa000df/srtcore/sync.cpp#L320)

### PR DESCRIPTION
**Pull request by ethouris**
Fixes #1104 

This adds error handling around sync resources. It's a shortcut for #1379 with only a fragment embracing the error handling.

### CODE REVIEW

**Comment by ethouris** [comment]
CodeFactor to ignore: Complex code in api.cpp.

File: srtcore/sync.cpp [hunk 1]
```diff
@@ -236,7 +236,11 @@ std::string srt::sync::FormatTimeSys(const steady_clock::time_point& timestamp)
 #if !defined(ENABLE_STDCXX_SYNC)
 srt::sync::Mutex::Mutex()
 {
-    pthread_mutex_init(&m_mutex, NULL);
+    int err = pthread_mutex_init(&m_mutex, 0);
```
Comment (maxsharabayko) [inline srtcore/sync.cpp:239]:
```suggestion
    const int err = pthread_mutex_init(&m_mutex, 0);
```

File: srtcore/core.cpp [hunk 2]
```diff
@@ -3582,7 +3582,16 @@ SRTSOCKET CUDT::makeMePeerOf(SRTSOCKET peergroup, SRT_GROUP_TYPE gtp, uint32_t l
     }
     else
     {
-        gp = &newGroup(gtp);
+        try
+        {
+            gp = &newGroup(gtp);
+        }
+        catch (...)
```
Comment (maxsharabayko) [inline srtcore/core.cpp:3589]: Waiting for any certain exception or any?

File: srtcore/api.cpp [hunk 3]
```diff
@@ -2598,6 +2598,7 @@ void CUDTUnited::updateMux(
        sockaddr_any sa;
        m.m_pChannel->getSockAddr((sa));
        m.m_iPort = sa.hport();
+   s->m_SelfAddr = sa; // Will be also completed later, but here it's needed for later checks
```
Comment (maxsharabayko) [inline srtcore/api.cpp]: There is something wrong with the alignment here.

File: srtcore/core.cpp (hunk 2 above)
Comment (ethouris) [reply srtcore/core.cpp:3614]: Probably only system exceptions and those freshly added here (although "added" is only in order to comply with the C++ std version).

File: srtcore/api.cpp [hunk 4]
```diff
@@ -2827,7 +2843,7 @@ SRTSOCKET CUDT::createGroup(SRT_GROUP_TYPE gt)
     {
         return APIError(e);
     }
-    catch (const std::bad_alloc& e)
+    catch (...)
```
Comment (maxsharabayko) [inline srtcore/api.cpp:2846]: Why `bad_alock` was not enough there?

Comment (ethouris) [reply srtcore/api.cpp:2846]: There could be some other types of exceptions thrown from the C++ standard library, as I remember. All of them qualify as IPE or at best lethal state of the system resources.

**Pull request merged**